Mise à jour du système de construction du code JavaScript

Alexinfos opened the pull request:
<!-- Décrivez vos changements ici (optionnel pour les tous petits changements). -->
Remplacement de Gulp par le bundler Vite.

Étapes à suivre :
- [x] Compilation liée au CSS
- [x] Production d'un sprite avec les icônes du site
- [x] Minification des images
- [x] Fichiers de police d'icônes
- [x] Transpilation du Javascript
- [x] Linting et formatting du Javascript
- [x] Modification du Makefile pour utiliser Vite
- [x] Retrait de Gulp

<!-- Remplacez XXXX par le numéro de ticket corrigé par vos changements : GitHub fermera le ticket mentionné automatiquement (voir https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue pour plus de détails). Supprimez la ligne s'il n'y a pas de ticket associé. --!>
Fix #6447

<!-- Si votre PR modifie du code Python, mettez à jour ou créez les tests unitaires associés. Signalez vos éventuelles difficultés afin que des contributeurs expérimentés puissent vous aider. -->

<!-- Si votre pull request requiert des actions particulières lors de la mise en production, renseignez-les ici afin qu’elles soient ajoutées au changelog lors du merge. -->

### Contrôle qualité

<!-- Donnez des instructions pour nous aider à vérifier vos changements.

Par exemple :

  - Lancez `python manage.py migrate` et `yarn test` ;
  - Créez un nouveau compte nommé `toto` ;
  - Envoyez un message privé à quelqu’un d’autre, le titre du message doit apparaître en rose. -->

Instructions données [plus loin dans la PR](https://github.com/zestedesavoir/zds-site/pull/6718#issuecomment-2823859829)

<!-- Merci ! -->

Alexinfos committed:
Installation de Vite.js et début de configuration pour bundler le CSS

Pour générer le CSS, lancez la commande `yarn vite build`.

Rq. : Les sprites n'étant pas encore générés par Vite, le fichier `_sprites.css`
est manquant, empêchant la compilation du SCSS. Un fix temporaire est de générer
ce fichier en utilisant Gulp (make build-front) et de copier `dist/scss/_sprite.scss`
dans `assets/scss/_sprite.scss`.

Alexinfos committed:
Remplacement des directives @import dépréciées par @use et @forward

A certains endroits, les directives import ont été gardées, une
bascule vers use/forward demandant des modifications majeures dans
l'organisation du code.

philippemilink commented:
Pourquoi est-ce qu'il est nécessaire de transformer tous les `@import` en `@use` / `@forward` dans les fichiers SCSS ?

Convolutio committed:
Ajout de spritesmith et build partiel avec vite

Import des css, mais mauvais placement pour le style
des pages d'erreur.
Script dynamique pour conserver la hiérarchie des fichiers
d'assets
Import (incomplet) des images d'assets
Manque les imports des assets en js

Alexinfos committed:
Ajout de la compression des images avec imagemin

Convolutio committed:
fix: explicit declaration of loaded assets

Convolutio committed:
Déclaration explicite des assets chargés par Vite

Convolutio committed:
Construction de tous les assets d'image par Vite

Alexinfos committed:
Migration de tous les fichiers JS vers des modules ECMAScript

Alexinfos committed:
Correction de l'emplacement du bundle js/script.js

Convolutio committed:
Copy statique des assets en js par ViteJS

TODO: faire en sorte que les assets en trop récupérés
de mathjax soient retirés du dist

Convolutio committed:
Moins de fichiers en trop de mathjax dans dist

Convolutio committed:
Build avec vite dans le Makefile

Convolutio committed:
Renommage fichier errors.css et nettoyage dist

Alexinfos committed:
Correction des chemins de certaines images

Convolutio commented:
## Contrôle qualité

1. Sur la branche `dev` de zestedesavoir, lancer une compilation avec Gulp.
    ```sh
    make build-front
    ```
    Ensuite, renommer le fichier `dist` en `correct-dist`
    ```sh
    mv dist correct-dist
    ```

2. Aller sur la branche associée à cette PR et mettre à jour / installer les nouveaux modules node avec
    ```sh
    yarn install
    ```
3. Lancer une construction avec ViteJS avec la même règle du Makefile
    ```sh
    make build-front
    ```
4. Comparer les fichiers produits entre les deux construction grâce à ce script sh
    ```sh
    # all useless diff data have been deleted
    # also ignore .map files in the comparison
    cd dist/ && find . | sort > ../dist.fd &&\
    cd ../correct-dist/ && find . | sort > ../correct-dist.fd &&\
    cd .. && diff dist.fd correct-dist.fd \
    | grep -E '^(<|>)' \
    | sed 's/>/Manquant :/g; s/</En trop :/g' \
    | sed '/\.map$/d' \
    > build_comparison.diff &&\
    rm dist.fd correct-dist.fd
    ```
5. Inspecter le fichier `build_comparison.diff` produit par ce script. Normalement, seul des fichiers `.woff` et `.woff2` non-utilisés par le css sont manquants. Des dossiers inutilisés par Django ont également été retirés, comme `images/sprite/` et `scss/`.  
    Ce script ne compare pas le dossier de sortie `errors/css/` mais il est possible de vérifier que le fichier `errors/css/main.css` existe bien.
    Les fichiers de sourcemap ne sont pas encore produits mais ne sont pas nécessaires au bon service du Frontend par Django.
    
6. Enfin, lancer Django pour servir le frontend
    ```sh
    source zdsenv/bin/activate
    make run-back
    ```
7. Vérifier dans l'inspecteur de votre navigateur sur `localhost:8000` que toutes les ressources sont chargées correctement

### Vérification du remplacement des autres fonctionnalités de Gulp

Avec les dépendances de ViteJS installées,

1. S'assurer que le watch mode fonctionne en lançant `make watch-front` et en éditant un fichier de frontend (e.g. `assets/scss/main.scss` après un premier build. Normalement, un second (et peut-être un troisième) build devrait se lancer.

2. S'assurer que le linting est faisable avec la règle `make lint-front`

3. S'assurer que la commande de réparation automatique de certaines règles de linting fonctionne avec la règle `make format-front`

## À corriger

- [x] ~~Pour le moment, le css produit ne résout pas correctement les chemins des images d'assets.~~ *(fixé !)*

## Remarques

- De nombreux warnings sur le css dûs aux dernières versions des préprocesseurs scss intégrés dans ViteJS devraient être affichés. Ils n'empêchent pas le bon déroulé de la construction du site de zds et peuvent donc être fixés à l'occasion d'une autre PR.
- De la documentation pour l'usage des url d'images dans les fichiers scss mériterait d'être rédigée pour que les développeurs puissent éditer du scss de manière adaptée au nouvel outil.
- Une construction par vitejs est encore assez lente à cause de la copie statique de librairies js et de fichiers de police. Ceci pourra être amélioré lorsqu'une intégration de vitejs avec Django pourra être mise en place. Il existe des plugins pour faire cela et le watch mode des fichiers de frontend pour alors être intégré dans un mode de développement de django (e.g. avec `make run-back`) et pourra bénéficier de 2 fonctionnalités clés de ViteJS permettant une accélération nette de la construction du frontend :
   - le pre-bundling des dépendances installées avec yarn et utilisées par les templates de django et les assets js 
   - la fonctionnalité du *Hot Module Replacement* de vitejs accélère énormément la construction pendant une phase de développement avec watch mode

Convolutio committed:
Chemins d'assets resolus dans les css par vitejs

En supplément, ajustement du vite.config.mjs pour sortir des
bundles css et js avec imports relatifs entre les fichiers
pour un chargement correct des assets du dossier dist par django,
peu importe le préfixage fixé par ce dernier

Convolutio committed:
Nouvelle règle de yarn pour le linting

Cette règle dépend de la dernière version de eslint
et est exécutable indépendamment de gulp

TODO: intégrer le formatteur stylish à eslint

Convolutio committed:
Watch mode pour la construction avec vite

Exclusion des assets intermédiaires pour éviter une boucle infinie de
construction dûe au watch mode
Modification du package.json pour mettre à jour a règle yarn

Convolutio committed:
Correction de l'activation du watch mode de vite

Avec yarn run build (sans le mode watch), on s'assure que la
construction s'achève bien et qu'aucune surveillance des fichiers n'est
faite par vitejs
Retrait également dans le Makefile du mode speed qui n'est pas utilisé
par vitejs et qui n'est pas pris en compte par son CLI

Convolutio committed:
Retrait de Gulp du projet

philippemilink commented on vite-src/pictures.html:
À quoi sert ce fichier ?

Convolutio replied:
Pour que vitejs traite des assets d'images (et les mettent au final dans le `dist/`), ces assets doivent être chargés dans un fichier source html, car le moteur (faignant/optimisé par défaut) ne traite que les assets réellement utilisés dans le frontend final. Ici, nous avons créé un frontend artificiel étant donné que le vrai est celui généré par django à partir des fichiers de template et des données du backend.

Je suis d'accord qu'on a là de l'ingénierie qui n'est pas produit habituellement dans les projets de vitejs, surtout que les projets vitejs qui comprennent ensuite un backend comme django utilisent généralement des plugins d'intégration pour que la construction en développement ou en production par vite se fasse en même temps que les constructions faites par django depuis ses templates (dans ce cas là, pas besoin alors de faire de fichiers pictures.html artificiels pour centraliser le chargement de tous les assets : on charge directement les assets utiles dans les fichiers de template de django)

philippemilink commented on vite.config.mjs:
```diff
@@ -0,0 +1,256 @@
+      outDir: "./dist",
+      assetsDir: ".",
+      sourcemap: true,
+      emptyOutDir: true, // TODO: remove when Gulp is needed
```
On peut donc supprimer cette ligne maintenant qu'on n'a plus de Gulp ?

philippemilink commented on assets/scss/base/_content.scss:
```diff
@@ -457,9 +460,6 @@ samp {
 }
 
-// code
-@import "base/source-code";
```
J'ai l'impression que `_source-code.scss` n'est donc plus inclus nul part. C'est un oubli ?

philippemilink reviewed (changes requested):
Je viens jeter un œil au code. J'ai l'impression que les tâches qu'il restaient à faire d'après le message initial de cette PR sont maintenant faites ?

En revanche, il reste à corriger les erreurs rapportées par le linter pour que la CI puisse passer et mettre à jour la documentation pour expliquer globalement comment fonctionne la construction du front.

Convolutio committed:
Retrait erreurs de linting dans sources de vite

Retrait des variables non-utilisées dans les fichiers
javascript rédigés pour vitejs

Convolutio committed:
Retrait d'une ligne de configuration inutile

Avec le retrait de Gulp, une ligne de configuration
de ViteJS ne sert plus à rien. Elle a été retirée

Convolutio committed:
Remplacement des path.resolve par path.join

Meilleure utilisation de la librairie path dans node
pour un calcul des chemins compatible avec plus de
plateformes

Alexinfos committed:
Ajout de l'import manquant de base/source-code

Alexinfos committed:
Mise à jour de la documentation

Convolutio committed:
Chemin de vite.config.mjs absolu et conforme à mjs

Utilisation de chemins absolus résolus pour le déplacement
et la suppression de fichiers dans le `dist/` produit par
vitejs.

Ce commit a pour objectif d'essayer de fixer les problèmes
de manipulation du système de fichier dans la CI

Convolutio committed:
Resolution de zmd/node_modules et aliasing

Encore une reecriture des chemins pour esperer qu'ils sont
resolus dans n'importe quel environnement, incluant celui de la CI

Convolutio committed:
Install de zmarkdown dans CI build-front

Pour tester une construction du frontend avec toutes
les dépendances requises, zmardkown est aussi installé
dans la tâche build-front de la CI
